src/sql.rs

```diff
@@ -299,28 +303,47 @@ impl Sql {
         }
     }
 
+    /// Locks the write transactions mutex. We use DEFERRED transactions for more parallelism, but
+    /// then db writers should be serialised, otherwise a write transaction A can retry for an
+    /// indefinite time and then fail by timeout if other writers modified the db since A
+    /// started. So, we sacrifice some parallelism for doing less work and more fairness in
+    /// servicing transactions.
+    pub async fn write_lock(&self) -> MutexGuard<'_, ()> {
+        self.wlk.lock().await
+    }
+
     /// Allocates a connection and calls given function with the connection.
     ///
     /// Returns the result of the function.
-    pub async fn call<'a, F, R>(&'a self, function: F) -> Result<R>
+    pub async fn call<F, R>(&self, function: F) -> Result<R>
     where
-        F: 'a + FnOnce(&mut Connection) -> Result<R> + Send,
-        R: Send + 'static,
+        F: FnOnce(&mut Connection) -> Result<R>,
     {
         let lock = self.pool.read().await;
         let pool = lock.as_ref().context("no SQL connection")?;
         let mut conn = pool.get().await?;
-        let res = tokio::task::block_in_place(move || function(&mut conn))?;
-        Ok(res)
+        function(&mut conn)
```

Comment:
You're right. I removed it because thought that transactions don't wait each other now and should be fast enough, but anyway it's a disk IO and also we don't know what function does. Maybe it does sleep(). Btw, have many worker threads do we have for `block_in_place()`?

Reply:
> There is no explicit number of worker set, so according to [the documentation](https://docs.rs/tokio/latest/tokio/runtime/struct.Builder.html#method.worker_threads) it is either the number of cores or whatever `TOKIO_WORKER_THREADS`.

This is true for the tokio executor, i.e. this is the number of threads we have if we do _not_ use `block_in_place`.

(FWIW `TOKIO_WORKER_THREADS` would be an environment variable that could be set)

> **Btw, how many worker threads do we have for `block_in_place()`?**

**As many as necessary, up to 512.**

`block_in_place` will use `spawn_blocking` to move the current executor's worker onto another thread:

https://github.com/tokio-rs/tokio/blob/88b1eb54fb66461b9f3524f4b5316241a019279a/tokio/src/runtime/scheduler/multi_thread/worker.rs#L337-L343

and then call call the passed closure on the current thread. 

Sidenote: If you want to do it the other way round (move the passed closure onto another thread and let the executor's worker continue on the current thread), use `spawn_blocking` directly.

From `spawn_blocking`'s [documentation](https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html):

> Tokio will spawn more blocking threads when they are requested through this function until the upper limit configured on the [Builder](https://docs.rs/tokio/latest/tokio/runtime/struct.Builder.html) is reached. After reaching the upper limit, the tasks are put in a queue. The thread limit is very large by default, because spawn_blocking is often used for various kinds of IO operations that cannot be performed asynchronously. When you run CPU-bound code using spawn_blocking, you should keep this large upper limit in mind. When running many CPU-bound computations, a semaphore or some other synchronization primitive should be used to limit the number of computation executed in parallel. Specialized CPU-bound executors, such as [rayon](https://docs.rs/rayon), may also be a good fit.

The upper limit [currently is 512](https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/builder.rs#L245).

---

BTW, would be nice to benchmark the three options of just calling the function (and blocking the executor), using `spawn_blocking` and using `block_in_place` against each other.

Reply:
Used @link2xt 's recipe for benchmarking (also disabled doctests in Cargo.toml) and didn't notice much difference:
```
$ hyperfine "cargo test" # block_in_place()
Benchmark 1: cargo test
  Time (mean ± σ):     36.687 s ±  1.034 s    [User: 41.570 s, System: 10.669 s]
  Range (min … max):   35.403 s … 38.818 s    10 runs
 
$ hyperfine "cargo test" # call F synchronously in tokio async executor
Benchmark 1: cargo test
  Time (mean ± σ):     36.652 s ±  1.078 s    [User: 41.286 s, System: 10.436 s]
  Range (min … max):   35.604 s … 39.545 s    10 runs
```
I think one need a rotational disk + page cache disabled to notice anything. Also i have a common question -- do we really need Rust/tokio asyncs at all? It's just a client code, not a server that creates hundreds of thousands threads/coroutines. From my C xperience it is very difficult nowadays to write async code that outperforms Linux pthreads even in huge server loads. Yes, it's possible, but it is a question of some percent of performance really. NPTL implementation is very efficient, the main thing that slows down thread on Linux is KPTI (Meltdown mitigation), because it flushes TLB when switching to kernel mode, but it can be safely turned off in typical server setups. IDK how the things are on other platforms however.

Anyway, i'll return block_in_place() for the reasons stated above.

Reply:
Another one reason to use async code instead of threads on Linux is that AFAIR one thread consumes two pages of kernel memory (besides the stack of course). But it's also a question of some percent of total memory consuption usually

Reply:
There are other pros of asyncs however:
- Rust asyncs are lazy, i.e. no-ops if not awaited
- Not related to Rust, but in languages like C in async code you have exact points of execution threads switching thus making the code more resistant to data race bugs and making possible to write critical sections of code w/o mutexes (as if they are atomic), but only if your async executer is single-threaded

Reply:
Good question; I wasn't around when the decision to move to async was done (@dignifiedquire mainly drove it forward, see #1356). Obviously, now it would be quite an effort to go back to synchronous code; there are `await` statements everywhere, the rpc and python bindings kinda use async, we use async cancellation (`stop_io()` and usages of `race()`), and to me it seems that it wouldn't be worth it.

My gut feeling is that the thread switches _don't_ slow down things (and also your benchmarks suggest that they don't) - most of the time is spent in SQL queries, or in encrypting/decrypting/parsing.

I think that for Delta Chat the main con of async is the increased programming complexity (we need an executor, and we often can't use `unwrap_or_else` and friends), idk what the pros are.

Reply:
Finally, i think we can benefit from asyncs in performance (and battery usage) because of doing less user-kernel mode switches. IDK however how much KPTI affects ARM processors. But also can lose because of an extra offloading/synchronisation related to `block_in_place()`. Anyway, i don't think it's worth switching to threads back. It's just would be nice to benchmark somehow.

Reply:
Btw, one more cons of asyncs pointed by @flub: https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#which-kind-of-mutex-should-you-use
And i recalled one more: handling page faults. There were talks on adding userspace PF handlers, which are also useful for virtual machines, but i don't know the current status